data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java

```diff
@@ -3,300 +3,282 @@
-            authenticationType = AuthenticationType.AAD_ACCESS_TOKEN_PROVIDER;
+            authenticationType = AuthenticationType.AAD_ACCESS_TOKEN;
+        } else if (csb.getTokenProvider() != null) {
             tokenProvider = csb.getTokenProvider();
+            authenticationType = AuthenticationType.AAD_ACCESS_TOKEN_PROVIDER;
         } else {
-            authenticationType = AuthenticationType.AAD_DEVICE_LOGIN;
+            if (StringUtils.isNotBlank(csb.getUserUsername())) {
+                userUsername = csb.getUserUsername();
+            }
+            redirectUri = new URI("http://localhost");
+            authenticationType = AuthenticationType.AAD_USER_PROMPT;
         }
 
-        // Set the AAD Authority URI
-        String aadAuthorityId = (csb.getAuthorityId() == null ? DEFAULT_AAD_TENANT : csb.getAuthorityId());
+        aadAuthorityUrl = determineAadAuthorityUrl(csb.getAuthorityId());
+    }
+
+    private String determineAadAuthorityUrl(String aadAuthorityUrl) {
+        String aadAuthorityId = (aadAuthorityUrl == null ? ORGANIZATION_URI_SUFFIX : aadAuthorityUrl);
         String aadAuthorityFromEnv = System.getenv("AadAuthorityUri");
-        if (aadAuthorityFromEnv == null){
-            aadAuthorityUri = String.format("https://login.microsoftonline.com/%s", aadAuthorityId);
+        if (aadAuthorityFromEnv == null) {
+            aadAuthorityUrl = String.format("https://login.microsoftonline.com/%s", aadAuthorityId);
         } else {
-            aadAuthorityUri = String.format("%s%s%s", aadAuthorityFromEnv, aadAuthorityFromEnv.endsWith("/") ? "" : "/", aadAuthorityId);
+            aadAuthorityUrl = String.format("%s%s%s", aadAuthorityFromEnv, aadAuthorityFromEnv.endsWith("/") ? "" : "/", aadAuthorityId);
         }
+        return aadAuthorityUrl;
     }
 
-    String acquireAccessToken() throws DataServiceException, DataClientException {
+    protected String acquireAccessToken() throws DataServiceException, DataClientException {
         if (authenticationType == AuthenticationType.AAD_ACCESS_TOKEN) {
             return accessToken;
         }
 
         if (authenticationType == AuthenticationType.AAD_ACCESS_TOKEN_PROVIDER) {
             try {
-                return tokenProvider.call();
+                Future<String> future = executor.submit(tokenProvider);
+                return getAsyncResultNonblocking(future);
             } catch (Exception e) {
-                throw new DataClientException(clusterUrl, e.getMessage(), e);
+                throw new DataClientException(scope, e.getMessage(), e);
             }
         }
 
-        if (lastAuthenticationResult == null) {
-            acquireToken();
-        } else if (IsInvalidToken()) {
-            if (lastAuthenticationResult.getRefreshToken() == null) {
-                acquireToken();
-            } else {
-                lastAuthenticationResultLock.lock();
+        try {
+            lastClientApplicationLock.lock();
```

Comment:
MSAL does (though it has a known concurrency issue for large number of threads), but here we're using our own entity, lastAuthenticationResult, for faster operation by determining whether we already have a valid token, whether to request a refresh token, or whether to request a new one from scratch, For this we need to be thread-safe.

I agree that acquireNewAccessToken() can be removed from this locked section, and it would be no big deal - we'll sometimes get a fresh token when we didn't need to. But I don't see the benefit to that. In such a case, wouldn't it be cheaper for a waiting thread to wait for the first thread's external call to complete so that when its thread is executing again it'll then be able to trivially get the last token (which is very cheap)?
Since the current implementation is per-client, this seems to me to be the ideal approach.

Reply:
Not sure I follow.
lastXXX is your confidential or public client app, once it has been created, you rely on it for caching via acquire_token_silent.
What is there to lock, apart from the instantiation (which could be moved up instead of locking)?

Reply:
In the general case, we rely on the last[Conf|Public]ClientApp for caching.
But if that fails, we call acquireNewAccessToken() anew.

Also, as above, I don't understand the performance benefit of allowing multiple threads for the same client make remote calls (either to get a new token, or a refresh token) at the same time. By waiting, it seems to me we reduce the number of necessary external calls.

Reply:
Mind that 
a) once it's cached, the lock is just delaying the code
b) conceding the lock costs the loss of the CPU for a full cycle 
c) the case you are defending from (no cached token, and more then one thread wants a token) is rare at best

I don't think this is worth the complexity you add to the code

Reply:
a) once it's cached, the delay is trivial because no external calls need to be made
b) conceding the lock costs the loss of the CPU for a full cycle, but also saves us from having to do multiple external calls. The latter would be more.
c) I agree it would be an uncommon scenario, but I'm trying to be resilient.

I agree that removing the lock would simplify the code, though. If that's more important than the performance considerations above (which I agree wouldn't be manifested often), I'll remove the lock.